arviz/kde_utils.py

```diff
@@ -0,0 +1,812 @@
+    pdf = 1 / (2 * np.pi * ive(0, kappa)) * np.exp(np.cos(x - mu) - 1) ** kappa
+    return pdf
+
+
+def circular_mean(x):
+    sinr = np.sum(np.sin(x))
+    cosr = np.sum(np.cos(x))
+    mean = np.arctan2(sinr, cosr)
+    return mean
+
+
+def A1inv(x):
+    """
+    Inverse function of the ratio of the first and zeroth order
+    Bessel functions of the first kind.
+
+    Returns the value k, such that A1inv(x) = k, i.e. A1(k) = x.
+    """
+    if 0 <= x < 0.53:
+        return 2 * x + x ** 3 + (5 * x ** 5) / 6
+    elif x < 0.85:
+        return -0.4 + 1.39 * x + 0.43 / (1 - x)
+    else:
+        return 1 / (x ** 3 - 4 * x ** 2 + 3 * x)
+
+
+def kappa_mle(x):
+    mean = circular_mean(x)
+    kappa = A1inv(np.mean(np.cos(x - mean)))
+    return kappa
+
+
+def dct1d(x):
+    """
+    Discrete Cosine Transform in 1 Dimension
+
+    Parameters
+    ----------
+    x : numpy array
+        1 dimensional array of values for which the
+        DCT is desired
+
+    Returns
+    -------
+    output : DTC transformed values
+    """
+
+    x_len = len(x)
+
+    even_increasing = np.arange(0, x_len, 2)
+    odd_decreasing = np.arange(x_len - 1, 0, -2)
+
+    x = np.concatenate((x[even_increasing], x[odd_decreasing]))
+
+    w_1k = np.r_[
+        1,
+        (2 * np.exp(-(0 + 1j) * (np.arange(1, x_len)) * np.pi / (2 * x_len)))
+    ]
+    output = np.real(w_1k * fft(x))
+
+    return output
+
+
+def _fixed_point(t, N, k_sq, a_sq):
+    """
+    Implementation of the function t-zeta*gamma^[l](t) derived
+    from equation (30) in [1]
+
+    References
+    ----------
+    .. [1] Kernel density estimation via diffusion.
+       Z. I. Botev, J. F. Grotowski, and D. P. Kroese.
+       Ann. Statist. 38 (2010), no. 5, 2916--2957.
+    """
+
+    k_sq = np.asfarray(k_sq, dtype=np.float64)
+    a_sq = np.asfarray(a_sq, dtype=np.float64)
+
+    l = 7
+    f = np.sum(np.power(k_sq, l) * a_sq * np.exp(-k_sq * np.pi ** 2 * t))
+    f *= 0.5 * np.pi ** (2.0 * l)
+
+    for j in reversed(range(2, l)):
+        c1 = (1 + 0.5 ** (j + 0.5)) / 3
+        c2 = np.product(np.arange(1.0, 2 * j + 1, 2, dtype=np.float64))
+        c2 /= (np.pi / 2) ** 0.5
+        t_j = np.power((c1 * (c2 / (N * f))), (2.0 / (3.0 + 2.0 * j)))
+        f = np.sum(k_sq ** j * a_sq * np.exp(-k_sq * np.pi ** 2.0 * t_j))
+        f *= 0.5 * np.pi ** (2 * j)
+
+    out = t - (2 * N * np.pi ** 0.5 * f) ** (-0.4)
+    return out
+
+
+def _root(function, N, args, x):
+    # The idea here was borrowed from KDEpy implementation.
+    # The right bound is at most 0.01
+    found = 0
+    N = max(min(1050, N), 50)
+    tol = 10e-12 + 0.01 * (N - 50) / 1000
+
+    while found == 0:
+        try:
+            bw, res = brentq(function, 0, 0.01, args=args, full_output=True,
+                             disp=False)
+            found = 1 if res.converged else 0
+        except ValueError:
+            bw = 0
+            tol *= 2.0
+            found = 0
+        if bw <= 0:
```

Comment:
This could be combined with the next condition, right? `if bw <= 0 or tol  >= 0` or putted in other way is it important for the user to distinguish between these two convergence errors?

Reply:
Hmmm, I think I used two different warnings because I was interested in that. But if I think twice I see you are right and it is not something important to the user.